src/psyclone/domain/common/algorithm/psyir.py

```diff
@@ -61,6 +76,95 @@ def _validate_child(position, child):
         '''
         return isinstance(child, KernelFunctor)
 
+    def _def_routine_root_name(self, index):
+        '''Internal method that returns the proposed processed routine name
+        given the index of this invoke.
+
+        :param int index: the position of this invoke call relative to \
+            other invokes in the algorithm layer.
+
+        :returns: the proposed processed routine name for this invoke.
+        :rtype: str
+
+        '''
+        routine_root_name = "invoke_{0}".format(index)
+        if len(self.children) == 1:
+            # Add the name of the kernel if there is only one call
+            routine_root_name += "_" + self.children[0].name
+        return routine_root_name
+
+    def create_language_level_symbols(self, index):
```

Comment:
I may be missing some use case but I don't understand why this is a public method and takes an external `index`. 

This make the interface on all the tests be:
```
    psyir = create_alg_psyir(code)
    invoke = psyir.children[0]
    invoke.create_language_level_symbols(0)
    invoke.lower_to_language_level()
```
Where you must know the index that this invoke has in the algorithm code, disconnected after parsing the algorithm code. Also the lowering method fails if this hasn't been called first, this doesn't look very safe.

I think the indexing should be assigned inside `alg_trans.apply(psyir)` (as I remember we said that the number is given from the initial position of the invoke - when raising(before any Psyclone script transformation)) and should possibly be stored as a string (e.g passed as an string index parameter in `__init__` or in `create()`) and this be fixed for the live of the AlgorithmInvokeCall (*with some transformations like fusing invokes able to update it to "_index1_index2"). Although I don't mind if it is an integer for now until we do fuses.

Then the lowering method should directly call `self._create_language_level_root_name(self._index)` instead of possibly failing. The symbols shouldn't exist while they are not used (as they may not exist at all depending on the fusing operations).

Does that make sense?

Reply:
The problem is that the PSyclone Algorithm-layer is actually transformed into a processed algorithm-layer *and used to generate the PSy-layer*. When transforming (lowering) the PSyclone Algorithm-layer to processed Algorithm-layer we destroy the algorithm layer information that is required to generate the PSy-layer. This means we must generate the PSy-layer first (or copy everything). In order to generate the PSy-layer we need the `container` and `routine` symbols that are going to connect the PSy-layer and the processed algorithm layer. It is possible we could defer these but my working assumption is that they are needed at this point. These symbols are required for *both* the processed algorithm layer and the PSy-layer (as the former calls the latter). I would be very happy to discuss how and when we should create these symbols and where they should be stored.

The next question is when to create an index for an invoke (which is needed when generating the container and routine symbols). I was concerned about creating the indices as we raise the generic PSyIR to Algorithm-layer PSyIR as we would be relying on the processing happening in a particular order. I therefore deferred providing this value as well (and merged it with the creation of the required symbols). However, you are right that the order is determined in `alg_trans.apply()` so I should and will provide the index here instead.

The steps that I think are needed are below and I guess we need to agree on how to support step 3 - or agree on a different approach.

1. Create Algorithm-layer PSyIR, including invoke index
2. Transform/optimise Algorithm-layer PSyIR (merge invokes etc and update index as appropriate)
3. Generate Container and Routine symbols
4. Create PSy-layer PSyIR using Algorithm-layer PSyIR and Container and Routine symbols (+ kernel metadata)
5. Create Processed (lowered) Algorithm PSyIR from Algorithm-layer PSyIR using Container and Routine symbols
6. Transform/optimise Processed Algorithm PSyIR, PSy-layer PSyIR and Kernel PSyIR

Reply:
We agreed in the call that providing the index must be done in the constructor but there are alternatives about when exactly to create the symbols. This depends on the ordering of steps the PSyclone does. The main alternatives are: create in the Transformation and provide them in the constructor, or create them by a getter and store them (subsequent calls to the get return the already created), in both cases they need to be stored in a attribute inside by the class.

Reply:
Moved setting of the index to the constructor. Note, this could be changed to a list of integers in the future which would allow the capturing of the history of any invoke fusion.